### DISCUSSION TITLE
Get openings in walls

### DISCUSSION THREAD

**Discussion by jurvanoerle** [IfcOpenShell/IfcOpenShell#4470] [Q&A] [answered]
With settings.set(settings.DISABLE_OPENING_SUBTRACTIONS, True) I get walls in my model without the openings. 
I want to obtain the openings a separate objects, so I can deduce the dimensions/position of the openings. Is there a similar way for openings as I use for verts? Like the code below:
  shape = iterator.get()
  verts = shape.geometry.verts
  openings = shape.geometry.openings (this obviously does not work)
What is the correct way to get the openings?

BTW, Is there a complete course/manual for python programming of ifcOpensHell. I feel quite clumsy asking these beginners questions.

Regards, Jur

**Comment by Moult** [comment, marked as answer]
The reason there isn't "complete" documentation is because IfcOpenShell's API is largely generated from the IFC schema itself. So to understand all possible relationships and attributes and so on you need to read the IFC docs.

So for example in the IFC documentation they describe how openings use the HasOpenings attribute: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/concepts/Object_Composition/Element_Voiding/Element_Openings/content.html

We can apply this as follows to get openings:

```python
openings = [r.RelatedOpeningElement for r in wall.HasOpenings]
```

After that you can run the geometry iterator on those openings just like you would any element. Then you can also use ifcopenshell.util.shape to calculate things like width and height and so on.
